sql: Aggregate of an aggregate crashes instead of giving an error

A query like `SELECT AVG(VARIANCE(x)) FROM t;` will currently crash with a nil pointer deference. Aggregate functions should not be allowed to be nested, which would avoid this issue. Postgres returns `"ERROR:  aggregate function calls cannot be nested"` when given this query.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4786b01]

goroutine 341 [running]:
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Eval(0xc820452960, 0x1, 0xece4593d1, 0xf57b9e5, 0x5971560, 0xece4593d1, 0xf57b9e5, 0x5971560, 0xc8202b43d0, 0xc8205ff540, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1142 +0x541
github.com/cockroachdb/cockroach/sql.(*selectNode).renderRow(0xc820558000)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/select.go:481 +0x17f
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc820558000, 0x1)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/select.go:125 +0x1ef
github.com/cockroachdb/cockroach/sql.(*groupNode).computeAggregates(0xc820192000)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/group.go:246 +0x70
github.com/cockroachdb/cockroach/sql.(*groupNode).Next(0xc820192000, 0xc82018a540)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/group.go:233 +0x3b
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func1(0xece4593d1, 0xf57b9e5, 0x5971560, 0xc82065b601, 0x28)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:452 +0x36a
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func2(0xc82015c280, 0x1)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:497 +0x121
```

--- a/sql/group.go
+++ b/sql/group.go
@@ -404,8 +404,9 @@ type extractAggregatesVisitor struct {
 	groupStrs map[string]struct{}
 
 	// groupedCopy is nil when visitor is in an Expr subtree that appears in the GROUP BY clause.
-	groupedCopy *extractAggregatesVisitor
-	err         error
+	groupedCopy         *extractAggregatesVisitor
+	subAggregateVisitor isAggregateVisitor
+	err                 error
 }
 
 var _ parser.Visitor = &extractAggregatesVisitor{}
@@ -432,6 +433,13 @@ func (v *extractAggregatesVisitor) Visit(expr parser.Expr, pre bool) (parser.Vis
 				// if an aggregate function of the wrong arity gets here,
 				// something has gone really wrong.
 				panic(fmt.Sprintf("%s has %d arguments (expected 1)", t.Name.Base, len(t.Exprs)))
+			}
+
+			defer v.subAggregateVisitor.reset()
+			_ = parser.WalkExpr(&v.subAggregateVisitor, t.Exprs[0])
+			if v.subAggregateVisitor.aggregated {
+				v.err = fmt.Errorf("aggregate function calls cannot be nested under %s", t.Name)
+				return v, expr
 			}
 
 			f := &aggregateFunc{

--- a/sql/testdata/aggregate
+++ b/sql/testdata/aggregate
@@ -184,6 +184,9 @@ SELECT COUNT(*), v/(k+v) FROM kv GROUP BY k+v
 query error aggregate functions are not allowed in WHERE
 SELECT k FROM kv WHERE AVG(k) > 1
 
+query error aggregate function calls cannot be nested under MAX
+SELECT MAX(AVG(k)) FROM kv
+
 # Test case from #2761.
 query II rowsort
 SELECT count(kv.k) AS count_1, kv.v + kv.w AS lx FROM kv GROUP BY kv.v + kv.w
@@ -206,13 +209,16 @@ SELECT UPPER(s), COUNT(DISTINCT s), COUNT(DISTINCT UPPER(s)) FROM kv GROUP BY UP
 A 2 1
 
 query II rowsort
-SELECT MAX(k), MIN(v) FROM kv HAVING MIN(v) > 2;
+SELECT MAX(k), MIN(v) FROM kv HAVING MIN(v) > 2
 ----
 
 query II rowsort
-SELECT MAX(k), MIN(v) FROM kv HAVING MAX(v) > 2;
+SELECT MAX(k), MIN(v) FROM kv HAVING MAX(v) > 2
 ----
 8 2
+
+query error aggregate function calls cannot be nested under MAX
+SELECT MAX(k), MIN(v) FROM kv HAVING MAX(MIN(v)) > 2
 
 query error argument of HAVING must be type bool, not type int
 SELECT MAX(k), MIN(v) FROM kv HAVING k;

